File: windows/src/AirClip.Net/HandshakeCodec.cs
using System.Buffers.Binary;
using System.Text;
using AirClip.Crypto;

namespace AirClip.Net;

public enum HandshakeFrameType : byte
{
    Hello = 1,
    Proof = 2,
    Reject = 3,
}

/// <summary>Who is on the other end, and the random half of the session key they contributed.</summary>
public sealed record HandshakeHello(
    string DeviceId, string DeviceName, string Platform, string Fingerprint, byte[] Challenge);

/// <summary>
/// The handshake's own binary frames, sent before any clipboard traffic. They are deliberately <em>not</em>
/// <c>ClipMessage</c> JSON: the protocol's message schema is fixed and has no room for a challenge or a
/// MAC, and bending <c>type</c> or <c>payload</c> to carry them would make AirClip's wire format subtly
/// different from the one it documents. A binary frame on the same socket keeps the JSON schema exact.
/// </summary>
public static class HandshakeCodec
{
    public const byte Version = 1;
    private const int MaxStringBytes = 256;
    private static readonly byte[] Magic = "ACLP"u8.ToArray();

    public static byte[] WriteHello(HandshakeHello hello)
    {
        ArgumentNullException.ThrowIfNull(hello);
        var frame = new List<byte>(160);
        frame.AddRange(Magic);
        frame.Add(Version);
        frame.Add((byte)HandshakeFrameType.Hello);
        WriteString(frame, hello.DeviceId);
        WriteString(frame, hello.DeviceName);
        WriteString(frame, hello.Platform);
        WriteString(frame, hello.Fingerprint);
        frame.AddRange(hello.Challenge);
        return [.. frame];
    }

    public static byte[] WriteProof(ReadOnlySpan<byte> mac)
    {
        var frame = new List<byte>(mac.Length + 7);
        frame.AddRange(Magic);
        frame.Add(Version);
        frame.Add((byte)HandshakeFrameType.Proof);
        frame.Add((byte)mac.Length);
        frame.AddRange(mac);
        return [.. frame];
    }

    public static byte[] WriteReject(string reason)
    {
        var frame = new List<byte>(64);
        frame.AddRange(Magic);
        frame.Add(Version);
        frame.Add((byte)HandshakeFrameType.Reject);
        WriteString(frame, reason);
        return [.. frame];
    }

    public static bool TryReadType(ReadOnlySpan<byte> frame, out HandshakeFrameType type)
    {
        type = default;
        if (frame.Length < 6 || !frame[..Magic.Length].SequenceEqual(Magic) || frame[4] != Version)
        {
            return false;
        }

        type = (HandshakeFrameType)frame[5];
        return true;
    }

    public static bool TryReadHello(ReadOnlySpan<byte> frame, out HandshakeHello? hello)
    {
        hello = null;
        if (!TryReadType(frame, out HandshakeFrameType type) || type != HandshakeFrameType.Hello)
        {
            return false;
        }

        int cursor = 6;
        if (!TryReadString(frame, ref cursor, out string? id)
            || !TryReadString(frame, ref cursor, out string? name)
            || !TryReadString(frame, ref cursor, out string? platform)
            || !TryReadString(frame, ref cursor, out string? fingerprint)
            || frame.Length - cursor != SessionCrypto.ChallengeSize)
        {
            return false;
        }

        hello = new HandshakeHello(id!, name!, platform!, fingerprint!, frame[cursor..].ToArray());
        return true;
    }

    public static bool TryReadProof(ReadOnlySpan<byte> frame, out byte[]? mac)
    {
        mac = null;
        if (!TryReadType(frame, out HandshakeFrameType type) || type != HandshakeFrameType.Proof
            || frame.Length < 7)
        {
            return false;
        }

        int length = frame[6];
        if (frame.Length - 7 != length)
        {
            return false;
        }

        mac = frame[7..].ToArray();
        return true;
    }

    public static bool TryReadReject(ReadOnlySpan<byte> frame, out string? reason)
    {
        reason = null;
        if (!TryReadType(frame, out HandshakeFrameType type) || type != HandshakeFrameType.Reject)
        {
            return false;
        }

        int cursor = 6;
        return TryReadString(frame, ref cursor, out reason);
    }

    private static void WriteString(List<byte> frame, string? value)
    {
        byte[] bytes = Encoding.UTF8.GetBytes(value ?? string.Empty);
        if (bytes.Length > MaxStringBytes)
        {
            // Trimmed back to a character boundary rather than to the byte limit: a device called
            // "办公室…" spends three bytes per character, and cutting one in half would put an invalid
            // UTF-8 sequence on the wire for the far end to render as a replacement glyph.
            int limit = MaxStringBytes;
            while (limit > 0 && (bytes[limit] & 0xC0) == 0x80)
            {
                limit--;
            }

            bytes = bytes[..limit];
        }

        Span<byte> length = stackalloc byte[sizeof(ushort)];
        BinaryPrimitives.WriteUInt16BigEndian(length, (ushort)bytes.Length);
        frame.AddRange(length);
        frame.AddRange(bytes);
    }

    private static bool TryReadString(ReadOnlySpan<byte> frame, ref int cursor, out string? value)
    {
        value = null;
        if (frame.Length - cursor < sizeof(ushort))
        {
            return false;
        }

        int length = BinaryPrimitives.ReadUInt16BigEndian(frame[cursor..]);
        cursor += sizeof(ushort);
        if (length > MaxStringBytes || frame.Length - cursor < length)
        {
            return false;
        }

        value = Encoding.UTF8.GetString(frame.Slice(cursor, length));
        cursor += length;
        return true;
    }
}

File: android/app/src/main/kotlin/com/airclip/core/protocol/HandshakeCodec.kt
package com.airclip.core.protocol

import com.airclip.core.crypto.SessionCrypto
import java.io.ByteArrayOutputStream

enum class HandshakeFrameType(val id: Byte) {
    HELLO(1),
    PROOF(2),
    REJECT(3),
    ;

    companion object {
        fun from(id: Byte): HandshakeFrameType? = values().firstOrNull { it.id == id }
    }
}

/** Who is on the other end, and the random half of the session key they contributed. */
class HandshakeHello(
    val deviceId: String,
    val deviceName: String,
    val platform: String,
    val fingerprint: String,
    val challenge: ByteArray,
)

/**
 * The handshake's own binary frames, sent before any clipboard traffic. A port of
 * `AirClip.Net.HandshakeCodec`, and like it they are deliberately *not* [ClipMessage] JSON: the
 * protocol's message schema is fixed and has no room for a challenge or a MAC, and bending `type` or
 * `payload` to carry them would make AirClip's wire format subtly different from the one it documents.
 * A binary frame on the same socket keeps the JSON schema exact.
 *
 * ```
 * offset size field
 * 0      4    magic "ACLP"
 * 4      1    version = 1
 * 5      1    frame type
 * 6      …    body (see each writer)
 * ```
 */
object HandshakeCodec {
    const val VERSION: Byte = 1

    private const val HEADER_SIZE = 6
    private const val MAX_STRING_BYTES = 256
    private val MAGIC = "ACLP".toByteArray(Charsets.US_ASCII)

    /** Four length-prefixed strings, then the fixed-size challenge. */
    fun writeHello(hello: HandshakeHello): ByteArray {
        val frame = begin(HandshakeFrameType.HELLO)
        writeString(frame, hello.deviceId)
        writeString(frame, hello.deviceName)
        writeString(frame, hello.platform)
        writeString(frame, hello.fingerprint)
        frame.write(hello.challenge)
        return frame.toByteArray()
    }

    /** One byte of length, then the MAC: it is always 32 bytes, but the length keeps it self-describing. */
    fun writeProof(mac: ByteArray): ByteArray {
        val frame = begin(HandshakeFrameType.PROOF)
        frame.write(mac.size)
        frame.write(mac)
        return frame.toByteArray()
    }

    fun writeReject(reason: String): ByteArray {
        val frame = begin(HandshakeFrameType.REJECT)
        writeString(frame, reason)
        return frame.toByteArray()
    }

    fun readType(frame: ByteArray): HandshakeFrameType? {
        if (frame.size < HEADER_SIZE) return null
        for (i in MAGIC.indices) {
            if (frame[i] != MAGIC[i]) return null
        }
        if (frame[4] != VERSION) return null
        return HandshakeFrameType.from(frame[5])
    }

    fun readHello(frame: ByteArray): HandshakeHello? {
        if (readType(frame) != HandshakeFrameType.HELLO) return null

        val cursor = Cursor(HEADER_SIZE)
        val deviceId = readString(frame, cursor) ?: return null
        val deviceName = readString(frame, cursor) ?: return null
        val platform = readString(frame, cursor) ?: return null
        val fingerprint = readString(frame, cursor) ?: return null
        if (frame.size - cursor.at != SessionCrypto.CHALLENGE_SIZE) return null

        return HandshakeHello(
            deviceId = deviceId,
            deviceName = deviceName,
            platform = platform,
            fingerprint = fingerprint,
            challenge = frame.copyOfRange(cursor.at, frame.size),
        )
    }

    fun readProof(frame: ByteArray): ByteArray? {
        if (readType(frame) != HandshakeFrameType.PROOF || frame.size < HEADER_SIZE + 1) return null
        val length = frame[HEADER_SIZE].toInt() and 0xFF
        if (frame.size - (HEADER_SIZE + 1) != length) return null
        return frame.copyOfRange(HEADER_SIZE + 1, frame.size)
    }

    fun readReject(frame: ByteArray): String? {
        if (readType(frame) != HandshakeFrameType.REJECT) return null
        return readString(frame, Cursor(HEADER_SIZE))
    }

    private fun begin(type: HandshakeFrameType): ByteArrayOutputStream {
        val frame = ByteArrayOutputStream(160)
        frame.write(MAGIC)
        frame.write(VERSION.toInt())
        frame.write(type.id.toInt())
        return frame
    }

    private fun writeString(frame: ByteArrayOutputStream, value: String) {
        var bytes = value.toByteArray(Charsets.UTF_8)
        if (bytes.size > MAX_STRING_BYTES) {
            // Trimmed back to a character boundary rather than to the byte limit: a device called
            // "办公室…" spends three bytes per character, and cutting one in half would put an invalid
            // UTF-8 sequence on the wire for the far end to render as a replacement glyph.
            var limit = MAX_STRING_BYTES
            while (limit > 0 && (bytes[limit].toInt() and 0xC0) == 0x80) limit--
            bytes = bytes.copyOfRange(0, limit)
        }

        frame.write((bytes.size ushr 8) and 0xFF)
        frame.write(bytes.size and 0xFF)
        frame.write(bytes)
    }

    private fun readString(frame: ByteArray, cursor: Cursor): String? {
        if (frame.size - cursor.at < 2) return null
        val length = ((frame[cursor.at].toInt() and 0xFF) shl 8) or (frame[cursor.at + 1].toInt() and 0xFF)
        cursor.at += 2
        if (length > MAX_STRING_BYTES || frame.size - cursor.at < length) return null

        val value = String(frame, cursor.at, length, Charsets.UTF_8)
        cursor.at += length
        return value
    }

    /** A read position that survives being passed to a helper; C# spells this `ref int`. */
    private class Cursor(var at: Int)
}
